dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusPTX.java

```diff
@@ -0,0 +1,311 @@
+        return programBuilder.addScopedChild(mainThread, store);
+    }
+
+    @Override
+    public Object visitStoreRegister(LitmusPTXParser.StoreRegisterContext ctx) {
+        MemoryObject object = programBuilder.getOrNewObject(ctx.location().getText());
+        Register register = programBuilder.getOrCreateRegister(mainThread, ctx.register().getText(), getArchPrecision());
+        String sem = ctx.sem().content;
+        String scope;
+        if (sem.equals(Tag.PTX.WEAK)) {
+            if (ctx.scope() != null) {
+                throw new ParsingException("Weak store instruction doesn't need scope: " + ctx.scope().content);
+            }
+            scope = Tag.PTX.SYS;
+        } else if (sem.equals(Tag.PTX.REL) || sem.equals(Tag.PTX.RLX)) {
+            scope = ctx.scope().content;
+        } else {
+            throw new ParsingException("Store instruction doesn't support sem: " + ctx.sem().content);
+        }
+        Store store = EventFactory.PTX.newTaggedStore(object, register, sem, scope);
+        store.addFilters(ctx.store().storeProxy);
+        return programBuilder.addScopedChild(mainThread, store);
+    }
+
+    @Override
+    public Object visitLoadConstant(LitmusPTXParser.LoadConstantContext ctx) {
+        Register register = programBuilder.getOrCreateRegister(mainThread, ctx.register().getText(), getArchPrecision());
+        IValue constant = new IValue(new BigInteger(ctx.constant().getText()), getArchPrecision());
+        String sem = ctx.sem().content;
+        String scope;
+        if (sem.equals(Tag.PTX.WEAK)) {
+            if (ctx.scope() != null) {
+                throw new ParsingException("Weak load instruction doesn't need scope: " + ctx.scope().content);
+            }
+            scope = Tag.PTX.SYS;
+        } else if (sem.equals(Tag.PTX.ACQ) || sem.equals(Tag.PTX.RLX)) {
+            scope = ctx.scope().content;
+        } else {
+            throw new ParsingException("Load instruction doesn't support sem: " + ctx.sem().content);
+        }
+        return programBuilder.addChild(mainThread, EventFactory.newLocal(register, constant));
+    }
+
+    @Override
+    public Object visitLoadLocation(LitmusPTXParser.LoadLocationContext ctx) {
+        Register register = programBuilder.getOrCreateRegister(mainThread, ctx.register().getText(), getArchPrecision());
+        MemoryObject location = programBuilder.getOrNewObject(ctx.location().getText());
+        String sem = ctx.sem().content;
+        String scope;
+        if (sem.equals(Tag.PTX.WEAK)) {
+            if (ctx.scope() != null) {
+                throw new ParsingException("Weak load instruction doesn't need scope: " + ctx.scope().content);
+            }
+            scope = Tag.PTX.SYS;
+        } else if (sem.equals(Tag.PTX.ACQ) || sem.equals(Tag.PTX.RLX)) {
+            scope = ctx.scope().content;
+        } else {
+            throw new ParsingException("Load instruction doesn't support sem: " + ctx.sem().content);
+        }
```

Comment:
I think it's not worth to formalize the logic as a function because the validate tags of instructions are different. For example: `st`{`.weak`, `.rlx`, `.rel`}, `ld`{`.weak`, `.rlx`, `.acq`}, `atom`{`.acq_rel`, `.rlx`}...